### PR TITLE
feat: rewrite abstract fragments for grpc

### PR DESCRIPTION
**Pull request by ysmolski**
I have used DataSourcePlanningBehavior for the flag that enforces rewriting of fragments.
I had to refactor the code so that structure is being specified in the Factory's method instead of Planner.

Access to this data happens via DataSource.PlanningBehavior()

BEGIN_COMMIT_OVERRIDE
feat: always rewrite abstract fragments for grpc
END_COMMIT_OVERRIDE

## Summary by CodeRabbit

- **New Features**
  - Per-data-source planning behavior now comes from the factory and adds AlwaysFlattenFragments (auto-enabled for gRPC-backed sources).

- **Bug Fixes**
  - More consistent selection rewriting for interfaces/unions and __typename handling; improved error handling and fewer edge-case failures.

- **Documentation**
  - Clarified planning-behavior options and renamed typename-related flag for clearer intent.

- **Tests**
  - Expanded and refactored tests covering abstract-type rewriting, gRPC scenarios, and planning-behavior variants.

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.

### CODE REVIEW

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 1]
```diff
@@ -1825,6 +1817,18 @@ func (f *Factory[T]) UpstreamSchema(dataSourceConfig plan.DataSourceConfiguratio
...
 }
 
+func (f *Factory[T]) PlanningBehavior() plan.DataSourcePlanningBehavior {
+	b := plan.DataSourcePlanningBehavior{
+		MergeAliasedRootNodes:      true,
+		OverrideFieldPathFromAlias: true,
+		IncludeTypeNameFields:      true,
+	}
+	if f.grpcClient != nil || f.grpcClientProvider != nil {
+		b.OperationEnforceRewritingFragmentSelections = true
```
Comment (devsergiy) [inline v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: it could be just

```go
{
  OperationEnforceRewritingFragmentSelections: f.grpcClient != nil || f.grpcClientProvider != nil
}
```

File: v2/pkg/engine/plan/abstract_selection_rewriter.go [hunk 2]
```diff
@@ -64,7 +64,8 @@ type fieldSelectionRewriter struct {
 	upstreamDefinition *ast.Document
 	dsConfiguration    DataSource
 
-	skipFieldRefs []int
+	skipFieldRefs    []int
+	enforceRewriting bool
```
Comment (devsergiy) [inline v2/pkg/engine/plan/abstract_selection_rewriter.go]: What do you think about naming it like `alwaysFlattenFragments` or just `alwaysRewrite`

File: v2/pkg/engine/plan/abstract_selection_rewriter.go [hunk 3]
```diff
@@ -169,6 +175,13 @@ func (r *fieldSelectionRewriter) processUnionSelection(fieldRef int, unionDefRef
 }
 
 func (r *fieldSelectionRewriter) unionFieldSelectionNeedsRewrite(selectionSetInfo selectionSetInfo, unionTypeNames, entityNames []string) (needRewrite bool) {
+	if r.enforceRewriting {
```
Comment (devsergiy) [inline v2/pkg/engine/plan/abstract_selection_rewriter.go]: probably we could extract this checks to the method

File: v2/pkg/engine/plan/abstract_selection_rewriter.go [hunk 4]
```diff
@@ -74,11 +75,16 @@ type RewriteResult struct {
...
+		definition:         definition,
+		upstreamDefinition: upstreamDefinition,
+		dsConfiguration:    dsConfiguration,
 	}
+	behavior := r.dsConfiguration.PlanningBehavior()
+	r.enforceRewriting = behavior.OperationEnforceRewritingFragmentSelections
+	return &r
 }
 
 func (r *fieldSelectionRewriter) SetUpstreamDefinition(upstreamDefinition *ast.Document) {
```
Comment (devsergiy) [inline v2/pkg/engine/plan/abstract_selection_rewriter.go]: as `SetUpstreamDefinition` is not used anymore, we could remove it

File: v2/pkg/engine/plan/abstract_selection_rewriter.go [hunk 5]
```diff
@@ -74,11 +75,16 @@ type RewriteResult struct {
 
 var resultNotRewritten = RewriteResult{}
 
-func newFieldSelectionRewriter(operation *ast.Document, definition *ast.Document) *fieldSelectionRewriter {
-	return &fieldSelectionRewriter{
-		operation:  operation,
-		definition: definition,
+func newFieldSelectionRewriter(operation *ast.Document, definition *ast.Document, upstreamDefinition *ast.Document, dsConfiguration DataSource) *fieldSelectionRewriter {
```
Comment (devsergiy) [inline v2/pkg/engine/plan/abstract_selection_rewriter.go]: we could skip passing upstreamDefinition from arguments and get it from ds configuration

File: v2/pkg/engine/plan/abstract_selection_rewriter_test.go [hunk 6]
```diff
@@ -46,12 +46,17 @@ func TestInterfaceSelectionRewriter_RewriteOperation(t *testing.T) {
 				break
 			}
 		}
+		require.NotEqual(t, ast.InvalidRef, fieldRef)
+
+		// Schema parses just the schema without merging with the base schema.
+		// In fact, rewriter does not use def from the FakeFactory.
+		ds := testCase.dsBuilder.
```
Comment (devsergiy) [inline v2/pkg/engine/plan/abstract_selection_rewriter_test.go:49]: to not reinit ds here, we could use schema in each test case so that rewriter will start getting it from ds configuration

Comment (devsergiy) [reply v2/pkg/engine/plan/abstract_selection_rewriter_test.go:49]: so we could use ds again instead of builder in test case struct

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 7]
```diff
@@ -14,33 +14,39 @@ import (
...
 	// This method serves as a flag that factory should have a context
 	Context() context.Context
+
 	UpstreamSchema(dataSourceConfig DataSourceConfiguration[DataSourceSpecificConfiguration]) (*ast.Document, bool)
+	PlanningBehavior() DataSourcePlanningBehavior
 }
 
 type DataSourceMetadata struct {
-	// FederationMetaData - describes the behavior of the DataSource in the context of the Federation
+	// FederationMetaData describes the behavior of the DataSource in the context of the Federation.
```
Comment (devsergiy) [inline v2/pkg/engine/plan/datasource_configuration.go]: probably makes sense to change it to:
Federation specific configuration for keys, requires, provides directives and entity interfaces

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 8]
```diff
@@ -363,18 +372,21 @@ func (d *DirectiveConfigurations) RenameTypeNameOnMatchBytes(directiveName []byt
 	return directiveName
 }
 
+// DataSourcePlanningBehavior is a way to configure planning per DataSource statically.
```
Comment (devsergiy) [inline v2/pkg/engine/plan/datasource_configuration.go]: maybe smth like `represents datasource specific planning options` ?

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 9]
```diff
@@ -383,8 +395,13 @@ type DataSourcePlanningBehavior struct {
 	// When true expected response will be { "rootField": ..., "alias": ... }
 	// When false expected response will be { "rootField": ..., "original": ... }
 	OverrideFieldPathFromAlias bool
-	// IncludeTypeNameFields should be set to true if the planner allows to plan __typename fields
+
+	// IncludeTypeNameFields set to true will require the planner to plan __typename fields.
```
Comment (devsergiy) [inline v2/pkg/engine/plan/datasource_configuration.go]: not accurate, not require - is allows planner to take care of planning "__typename" fields

If this option set to false, planner will never plan __typename field

Comment (devsergiy) [reply v2/pkg/engine/plan/datasource_configuration.go]: probably we could rename it to smth like `AllowToPlanTypeNameField`

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 10]
```diff
@@ -383,8 +395,13 @@ type DataSourcePlanningBehavior struct {
...
 	IncludeTypeNameFields bool
+
+	// If true then planner will rewrite the operation
+	// to flatten inline fragments to only the concrete types.
+	OperationEnforceRewritingFragmentSelections bool
```
Comment (devsergiy) [inline v2/pkg/engine/plan/datasource_configuration.go]: maybe `AlwaysFlattenFragments` ?

File: v2/pkg/engine/plan/node_selection_visitor.go [hunk 11]
```diff
@@ -649,9 +649,7 @@ func (c *nodeSelectionVisitor) rewriteSelectionSetHavingAbstractFragments(fieldR
 		return
 	}
 
-	rewriter := newFieldSelectionRewriter(c.operation, c.definition)
-	rewriter.SetUpstreamDefinition(upstreamSchema)
-	rewriter.SetDatasourceConfiguration(ds)
+	rewriter := newFieldSelectionRewriter(c.operation, c.definition, upstreamSchema, ds)
```
Comment (devsergiy) [inline v2/pkg/engine/plan/node_selection_visitor.go]: per proposal to get upstream schema we could still check if it is exists on a datasource, but get it from ds

File: v2/pkg/engine/plan/path_builder_visitor.go [hunk 12]
```diff
@@ -1237,9 +1237,9 @@ func (c *pathBuilderVisitor) LeaveField(ref int) {
 	c.removeArrayField(ref)
 }
 
-// addPlannerPathForTypename adds a path for the __typename field
-// adding __typename should be done only in case particular planner has parent path
-// otherwise it will be added to all planners and will cause visiting of incorrect selection sets
+// addPlannerPathForTypename adds a path for the __typename field.
+// It should be called only if a particular planner has the parentPath,
```
Comment (devsergiy) [inline v2/pkg/engine/plan/path_builder_visitor.go]: not accurate
this sentence `adding __typename should be done only in case particular planner has parent path` describes behaviour inside a method, while we call it unconditionally

Comment (devsergiy) [reply v2/pkg/engine/plan/path_builder_visitor.go]: we could move the last 2 sentences to the beginning of method

**Review by devsergiy — CHANGES_REQUESTED** [review]

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go (hunk 1 above)
Comment (ysmolski) [reply v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: Thanks, you are right.

File: v2/pkg/engine/plan/abstract_selection_rewriter.go (hunk 2 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/abstract_selection_rewriter.go]: I like `alwaysRewrite`. Also I have renamed the flag in the Behavior to `AlwaysRewriteFragmentSelections`

File: v2/pkg/engine/plan/abstract_selection_rewriter.go (hunk 4 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/abstract_selection_rewriter.go]: ok

File: v2/pkg/engine/plan/abstract_selection_rewriter.go (hunk 5 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/abstract_selection_rewriter.go]: doing it

File: v2/pkg/engine/plan/abstract_selection_rewriter.go (hunk 3 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/abstract_selection_rewriter.go]: extracted

File: v2/pkg/engine/plan/datasource_configuration.go (hunk 7 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/datasource_configuration.go]: thanks

File: v2/pkg/engine/plan/datasource_configuration.go (hunk 8 above)
Comment (devsergiy) [reply v2/pkg/engine/plan/datasource_configuration.go]: hmm, represents additional planning configuration

Comment (ysmolski) [reply v2/pkg/engine/plan/datasource_configuration.go]: what about?
`// DataSourcePlanningBehavior contains DataSource-specific planning flags.`

Comment (ysmolski) [reply v2/pkg/engine/plan/datasource_configuration.go]: or even this structure could be renamed to 
`// DataSourcePlanningFlags contains DataSource-specific planning flags.`

No behavior here, just the flags.

File: v2/pkg/engine/plan/datasource_configuration.go (hunk 10 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/datasource_configuration.go]: Sounds right to me. 

"Naming things is the most hardest thing in computer science."

File: v2/pkg/engine/plan/datasource_configuration.go (hunk 9 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/datasource_configuration.go]: done

File: v2/pkg/engine/plan/path_builder_visitor.go (hunk 12 above)
Comment (ysmolski) [reply v2/pkg/engine/plan/path_builder_visitor.go]: My bad. I will make those sentences as internal comment.

**Review by devsergiy — APPROVED** [review]

File: v2/pkg/engine/plan/datasource_configuration.go (hunk 8 above)
Comment (devsergiy) [reply v2/pkg/engine/plan/datasource_configuration.go]: flags defines the behaviour so it is fine
